RUM-8415 Add SwiftUI view auto-instrumentation

mariedm opened the pull request:
### What and why?

A short description of what changes this PR introduces and why.

### How?

A brief description of implementation details of this PR.

### Review checklist
- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes
- [ ] Add Objective-C interface for public APIs (see our [guidelines](https://datadoghq.atlassian.net/wiki/spaces/RUMP/pages/3157787243/RFC+-+Modular+Objective-C+Interface#Recommended-solution) [internal]) and run `make api-surface`)

mariedm commented on DatadogRUM/Sources/Instrumentation/RUMInstrumentation.swift:
```diff
@@ -63,7 +63,9 @@ internal final class RUMInstrumentation: RUMCommandPublisher {
-            predicate: uiKitRUMViewsPredicate,
+            uiKitPredicate: uiKitRUMViewsPredicate,
+            swiftUIPredicate: nil,
+            swiftUIViewNameExtractor: nil,
```
Both `nil` for now, preventing customers from using.

mariedm commented on DatadogRUM/Sources/Instrumentation/Views/RUMView.swift:
```diff
@@ -0,0 +1,49 @@
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
```
No change here, just moved `RUMView` to its own file.

mariedm commented on DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/SwiftUI/SwiftUIWireframesBuilder.swift:
```diff
@@ -223,17 +223,4 @@ internal extension ImagePrivacyLevel {
     }
 }
 
-#if DEBUG
```
Moved to `CustomDump.swift`.

maxep commented on DatadogRUM/Sources/Instrumentation/Views/SwiftUI/SwiftUIViewNameExtractor+NavigationStackController.swift:
```diff
@@ -0,0 +1,103 @@
+    @usableFromInline
+    internal func extractNavigationControllerPath(reflector: ReflectorType) -> String? {
+        // List / Navigation Split detailed pattern
+        if let navStack = reflector.descendantIfPresent("host", "_rootView", "storage", "view", "content", "content", "content", "content", "list", "item", "type") {
```
/question How are these paths stable across iOS versions? I'm a bit concerned about these breaking without visibility in telemetry. Maybe we could split it with `"host", "_rootView", "storage", "view", "content"` first and then try to access subpaths.

mariedm replied:
Regarding telemetry, we can inject `core.telemetry` when creating the `reflectorFactory`. 

Is the idea of splitting the path to identify at which level the traversal might fail? In any case, it makes sense to break the traversal into smaller, incremental steps.

mariedm replied:
And yes, I'll use enum, it makes sense. 👍

mariedm replied:
I've logged a ticket to tackle telemetry separately. Given that some paths are ignored on purpose, this requires more thinking on what is considered an actual failure in path traversal.

maxep commented on DatadogRUM/Sources/Instrumentation/Views/SwiftUI/SwiftUIViewNameExtractor+NavigationStackController.swift:
```diff
@@ -0,0 +1,103 @@
+
+    // MARK: - String Parsing
+    @usableFromInline
+    internal func extractViewNameFromNavigationStackHostingController(_ input: String) -> String? {
```
/suggestion Despite the great comments, I have a hard time understanding the logic and it will be difficult to maintain... It think it would be best to use regex instead, WDYT?

mariedm replied:
I initially used a regex approach, but I found it difficult (impossible?) to handle all the use cases.

maxep commented on DatadogRUM/Sources/Instrumentation/Views/SwiftUI/SwiftUIViewNameExtractor.swift:
```diff
@@ -0,0 +1,140 @@
+    // MARK: - Helpers
+    internal func shouldSkipViewController(className: String, viewController: UIViewController) -> Bool {
+        // Skip Tab Bar Controllers as they're containers
+        if className == "SwiftUI.UIKitTabBarController" {
```
/question Isn't it a subclass of `UITabBarController`?

maxep reviewed (commented):
Looks good 👍 

I have done a first round with few questions/suggestions, LMKWYT!

simaoseica-dd reviewed (commented):
I've added some questions, mainly about using regex to extract the names from the view tree. 
This could improve readability and help enumerate the use cases we are tracking.

mariedm commented on DatadogRUM/Sources/Instrumentation/Views/SwiftUI/SwiftUIViewNameExtractor+HostingController.swift:
```diff
@@ -0,0 +1,65 @@
+
+extension SwiftUIReflectionBasedViewNameExtractor {
+    // MARK: - Path Traversal
+    @usableFromInline
```
I had added it because it seemed necessary for `SwiftUIReflectionBasedViewNameExtractor` to access it since it was in a different file (which seemed weird btw). I just tried to remove them and it's working now, so I'll get rid of them.

mariedm commented on DatadogInternal/Sources/Utils/Dumper.swift:
```diff
@@ -0,0 +1,31 @@
+///
+/// The `Dumper` protocol provides a standard interface for converting objects to string
+/// representations and writing them to a text output stream.
+public protocol Dumper {
```
The `Dumper` protocol allows dependency injection in the extractor. In the integration tests, we can inject the `TestDumper`, which returns a predetermined output.

mariedm replied:
Removing as we don't need this anymore since the extractor doesn't rely on `customDump` anymore.

mariedm commented on DatadogRUM/Sources/Instrumentation/Views/RUMViewsHandler.swift:
```diff
@@ -196,14 +209,29 @@ internal final class RUMViewsHandler {
+                  let rumView = swiftUIPredicate.rumView(for: rumViewName) {
             add(
                 view: .init(
                     identity: identity,
```
Oh yes, good catch!

mariedm commented on DatadogInternal/Sources/Utils/CustomDump.swift:
```diff
@@ -550,7 +548,18 @@ public struct FileHandlerOutputStream: TextOutputStream {
 
+#if DEBUG
+/// Dumps the given value's contents in a file.
+public func dump<T>(_ value: T, filename: String) throws {
```
Only for debug purposes.

mariedm commented on DatadogRUM/Sources/Instrumentation/Views/SwiftUI/SwiftUIViewNameExtractor.swift:
```diff
@@ -0,0 +1,140 @@
+
+    private func extractViewNameFrom(from viewController: UIViewController) -> String? {
+        // Reflector to inspect the view controller's internals
+        let reflector = createReflector(viewController)
```
Yes, good call

mariedm commented on DatadogRUM/Sources/Instrumentation/Views/SwiftUI/SwiftUIViewNameExtractor+NavigationStackController.swift:
```diff
@@ -0,0 +1,103 @@
+        // List / Navigation Split detailed pattern
+        if let navStack = reflector.descendantIfPresent("host", "_rootView", "storage", "view", "content", "content", "content", "content", "list", "item", "type") {
+            var output = ""
+            dumper.dump(navStack, to: &output)
```
Looks like I could use `String(describing: type(of: navStack))` instead. I'll update the logic accordingly.

mariedm commented:
@maxep @simaoseica-dd thank you for the thorough review! 

I've updated the name extraction logic:
- Replaced object dumping with type descriptions, which resulted in cleaner and more consistent inputs for name extraction
- Replaced string parsing with regex
- Made `ControllerType` initializable from class names
- Organized path traversal patterns into dedicated enums for better maintainability
- Updated tests accordingly
  -  Note: removed performance test since it's now less relevant, but will add one in upcoming UI tests

simaoseica-dd commented on DatadogRUM/Sources/Instrumentation/Views/SwiftUI/SwiftUIViewNameExtractor.swift:
```diff
@@ -0,0 +1,163 @@
+
+    // MARK: - Helpers
+    /// Extracts a view name from a type description using regex patterns
+    internal func extractViewName(from input: String) -> String? {
```
question: Can these helpers be private? I know the tests are calling them, but the access modifiers shouldn't be defined by the tests

mariedm replied:
What do you suggest instead? 

These helpers are only accessible within the RUM module and not exposed to customers. Using `@testable import` lets us test them while maintaining proper API encapsulation.

simaoseica-dd replied:
The suggestion would be to use `SwiftUIReflectionBasedViewNameExtractor` with the exposed `func extractName(from:` using initialized VCs.
As talked, lets address this in future iterations

simaoseica-dd commented on DatadogRUM/Sources/Instrumentation/Views/SwiftUI/SwiftUIViewPath.swift:
```diff
@@ -0,0 +1,54 @@
+import DatadogInternal
+
+/// Common path components for SwiftUI view traversal
+internal enum RootPath {
```
My suggestion was:

```swift
internal enum ViewNode: String {
    case host
    case rootView = "_rootView"
    
    static let hostingBase: [Self] = [.host, .rootView]
}
```

simaoseica-dd reviewed (approved):
Great work on the regex! 🎖️ 
Left some nitpicking, but overall looks great!

maxep commented on DatadogRUM/Sources/Instrumentation/Views/SwiftUI/SwiftUIViewNameExtractor.swift:
```diff
@@ -0,0 +1,159 @@
+    // MARK: - Helpers
+    /// Extracts a view name from a type description using regex patterns
+    internal func extractViewName(from input: String) -> String? {
+        // Pattern 1: Extract the view name from generic types like LazyView<ContentView>
+        let simpleGenericPattern = #"<([^<>,]+)>"#
+        if let match = input.range(of: simpleGenericPattern, options: .regularExpression) {
+            let startIndex = input.index(match.lowerBound, offsetBy: 1)
+            let endIndex = input.index(match.upperBound, offsetBy: -1)
+            return String(input[startIndex..<endIndex])
+        }
+
+        // Pattern 2: Extract the view name from complex generic types like ParameterizedLazyView<String, DetailView>
+        let complexGenericPattern = #"<.*,\s*([^<>,]+)>"#
+        if let match = input.range(of: complexGenericPattern, options: .regularExpression),
+           let captureRange = input.range(of: #"([^<>,]+)(?=>)"#, options: .regularExpression, range: match) {
+            return String(input[captureRange]).trimmingCharacters(in: .whitespaces)
+        }
```
/suggestion You can use a single regex for this using the `NSRegularExpression`:
```swift
let regex = try NSRegularExpression(pattern: #"<(?:[^,>]*,\s+)?([^<>,]+?)>"#, options: [])
if
    let match = regex.firstMatch(in: input, options: [], range: NSRange(input.startIndex..<input.endIndex, in: input)),
    let range = Range(match.range(at: 1), in: input)
{
    return String(input[range])
}
```

maxep commented on DatadogRUM/Tests/Instrumentation/Views/SwiftUIViewNameExtractorTests.swift:
```diff
@@ -0,0 +1,94 @@
+    // MARK: - View Name Extraction Tests
+    func testViewNameExtraction() {
+        let testCases: [(String, String)] = [
+            ("LazyView<ContentView>", "ContentView"),
+            ("SheetContent<Text>", "Text"),
+            ("Optional<Text>", "Text"),
+            ("Optional<ProfileView>", "ProfileView"),
+            ("LazyView<HomeView>", "HomeView"),
+            ("ParameterizedLazyView<String, DetailViewForNavigationDestination>", "DetailViewForNavigationDestination"),
+            ("DetailView.Type", "DetailView"),
+            ("SheetContent<ModalSheet>", "ModalSheet")
```
It's too simplistic, we should cover composed types such:
```swift
("ModifiedView<ModifierType, ModifiedView<ModifierType, ModifiedView<ModifierType, ModifiedView<ModifierType, ViewType>>>>", "ViewType"),
("ModifiedView<ModifierType, ModifiedView<ModifierType, ModifiedView<ModifierType, ModifiedView<ModifierType, ContainerType<ViewType>>>>>", "ViewType")
```

maxep commented on DatadogRUM/Sources/Instrumentation/Views/SwiftUI/SwiftUIViewNameExtractor.swift:
```diff
@@ -0,0 +1,159 @@
+internal struct SwiftUIReflectionBasedViewNameExtractor: SwiftUIViewNameExtractor {
+    private let createReflector: (Any) -> TopLevelReflector
+
+    init(
+        reflectorFactory: @escaping (Any) -> TopLevelReflector = { subject in
+            Reflector(
+                subject: subject,
+                // MARK: TODO: RUM-9035 - Telemetry
+                telemetry: NOPTelemetry()
+            )
+        }
+    ) {
+        self.createReflector = reflectorFactory
+    }
```
/nit It would read better if we have a `convenience init` that take a `telemetry` parameter.

maxep replied:
Sorry my comment was not clear, what I meant was to remove the default `reflectorFactory` value and add a `convenience init(telemetry: Telemetry = NOPTelemetry())` :)

maxep reviewed (commented):
Looks good, I think we can still improve the regex and its coverage in tests.

maxep commented on DatadogRUM/Sources/Instrumentation/Views/SwiftUI/SwiftUIViewNameExtractor.swift:
```diff
@@ -0,0 +1,172 @@
+
+    // MARK: - Helpers
+    private static let genericTypePattern: NSRegularExpression? = {
+        do {
+            return try NSRegularExpression(pattern: #"<(?:[^,>]*,\s+)?([^<>,]+?)>"#)
+        } catch {
+            return nil
+        }
```
/nit This is the same as `try? NSRegularExpression(pattern: #"<(?:[^,>]*,\s+)?([^<>,]+?)>"#)` 😉

mariedm replied:
Yep, true; at first I was logging a console message when failing but removed it, so it doesn't make sense to have the do catch anymore.

maxep commented on DatadogRUM/Sources/Instrumentation/Views/SwiftUI/TopLevelReflector.swift:
```diff
@@ -0,0 +1,19 @@
+}
+
+// MARK: - Reflector
+extension Reflector: TopLevelReflector {}
```
/question Do you really need `Reflector` to be compliant? I guess `ReflectionMirror` would be sufficient.

mariedm replied:
Now using `ReflectionMirror` instead of `Reflector` in the name extractor.

maxep reviewed (dismissed):
Looks great! Thanks for the updates. I left couple of minor/non-blocking suggestions!

simaoseica-dd reviewed (approved):
LGTM 👍

mariedm commented:
/merge
